src/common/client_response.hpp

```diff
@@ -0,0 +1,128 @@
+// every phase in dialog can have some buttons
+// buttons are generalized on this level as responses
+// because non GUI/WUI client can also use them
+
+#pragma once
+
+#include <cstdint>
+#include <cstddef>
+#include <array>
+enum { RESPONSE_BITS = 2,
+    MAX_RESPONSES = (1 << RESPONSE_BITS) };
+
+//list of all button types
+enum class Response : uint8_t {
+    _none = 0, //none must be zero becouse of empty initialization of array
+    Yes,
+    No,
+    Continue,
+    Ok,
+    Back,
+    Retry,
+    Purge_more,
+    Reheat
+};
+
+using PhaseResponses = std::array<Response, MAX_RESPONSES>;
+
+//count cenum class members (if "_first" and "_last" is defined)
+template <class T>
+constexpr size_t CountPhases() {
+    return static_cast<size_t>(T::_last) - static_cast<size_t>(T::_first) + 1;
+}
+//use this when creating an event
+//encodes enum to position in Enem
+template <class T>
+constexpr uint8_t GetPhaseIndex(T phase) {
+    return static_cast<size_t>(phase) - static_cast<size_t>(T::_first);
+}
+
+template <class T>
+constexpr T GetEnumFromPhaseIndex(size_t index) {
+    return static_cast<T>(static_cast<size_t>(T::_first) + index);
+}
+
+//define enum classes for responses here
+//and YES phase can have 0 responses
+//every enum must have "_first" and "_last"
+//"_first" ==  "previous_enum::_last" + 1
+//EVERY response shall have unique ID (so every button in GIU is unique)
+enum class PhasesLoadUnload : uint16_t {
+    _first = 0,
+    Parking,
+    WaitingTemp,
+    PreparingToRam,
+    Ramming,
+    Unloading,
+    Unloading2,
+    UserPush,
+    NozzleTimeout,
+    MakeSureInserted,
+    Inserting,
+    Loading,
+    Purging,
+    Purging2,
+    IsColor,
+    Unparking,
+    _last = Unparking
+};
+
+enum class PhasesTest : uint16_t {
+    _first = static_cast<uint16_t>(PhasesLoadUnload::_last) + 1,
+    Test1,
+    Test2,
+    _last = Test2
+};
+
+//static class for work with fsm responses (like button click)
+//encode responses - get them from marlin client, to marlin server and decode them again
+class ClientResponses {
+    ClientResponses() = delete;
+    ClientResponses(ClientResponses &) = delete;
+
+    //declare 2d arrays of single buttons for radio buttons
+    static const PhaseResponses LoadUnloadResponses[CountPhases<PhasesLoadUnload>()];
+    static const PhaseResponses TestResponses[CountPhases<PhasesTest>()];
+
+    //methods to "bind" button array with enum type
+    static const PhaseResponses &getResponsesInPhase(PhasesLoadUnload phase) { return LoadUnloadResponses[static_cast<size_t>(phase)]; }
+    static const PhaseResponses &getResponsesInPhase(PhasesTest phase) { return TestResponses[static_cast<size_t>(phase)]; }
+
+protected:
+    //get index of single response in PhaseResponses
+    template <class T>
+    static uint8_t GetIndex(T phase, Response response) {
```

Comment:
Return value not documented. Type of return value is unsigned, but we return negative number.

Reply:
documented